source/module_psi/psi_initializer.h

```diff
@@ -0,0 +1,267 @@
+        /// @return if GlobalV::init_wfc has valid value, return the number of wavefunctions to be initialized, otherwise throw an error
+        int get_starting_nw() const;
+
+        /// @brief allocate memory for psi
+        /// @return pointer to psi, memory allocated
+        psi::Psi<std::complex<double>>* allocate();
+
+        /// @brief calculate psi in planewave representation
+        /// @param psi psi
+        /// @param ik index of kpoint
+        virtual psi::Psi<std::complex<double>>* cal_psig(int ik) = 0;
+
+        /// @brief initialize planewave represented psi after change of cell volume
+        /// @note due to wanf2, the Wannier function in pw representation, is not used anymore, this function is not needed anymore
+        void init_after_vc(); // in the past this function will refresh the wanf2, the Wannier function in pw representation. However for now it is not needed, because lcao_in_pw now is designed for initial guess
+        
+        /// @brief get method of initializing psi
+        /// @return the method
+        std::string get_method() const { return this->method; }
+        /// @brief get complement number of bands
+        /// @return nbands_complem
+        int get_nbands_complem() const { return this->nbands_complem; }
+
+        void print_status(psi::Psi<std::complex<double>>& psi) const;
+
+        /// @brief set method manually
+        /// @param method_in initialization method
+        void set_method(std::string method_in) { this->method = method_in; }
+        /// @brief set number of complementary bands
+        /// @param nbands_in nbands_complem
+        void set_nbands_complem(int nbands_in) { this->nbands_complem = nbands_in; }
+        /// @brief output psig to file, given number of kpoints, bands and basis, diagonalization method. Note: because it is complex number, therefore every number has format (real, imag)
+        void write_psig() const;
+        /// @brief output psig at given kpoint to file
+        void write_psig(int ik) const;
+        // virtual functions, will be implemented in derived classes
+
+        // random to complement bands not initialized by pswfc or nao, therefore it is a basic function, or psi_initializer_random will be inherented by all other methods.
+        /// @brief kernel to generate and assign random number for psi
+        /// @param psi for psi::Psi<FPTYPE, Device>, first use psi.fix(ik), then psi.get_pointer() to pass to parameter list
+        /// @param iw_start index of wavefunction (start), the index of first band to initialize
+        /// @param iw_end index of wavefunction (end)
+        /// @param ik index of kpoint
+        /// @param wfc_basis because do not want to change contents in it, this parameter is reserved. should give this->pw_wfc
+        void random_t(std::complex<double>* psi, const int iw_start, const int iw_end, const int ik, const ModulePW::PW_Basis_K* wfc_basis)
```

Comment:
Please move this function to other .cpp

Reply:
For it will be call everytime when the atomic wavefunctions are not enough, I place it in psi_initializer.cpp. In previous it is a inline function with template parameters, however that will bring too many undefined reference troubles. Fortunately I find ESolver_FP::psi it self is a psi::Psi<std::complex<double>> type instance, so I change to only specialized it as double.